icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java

```diff
@@ -324,16 +330,20 @@ void buildTrie() {
     void mergeCategories(IntPair categories) {
         assert(categories.first >= 1);
         assert(categories.second > categories.first);
+        assert((categories.first <  fDictCategoriesStart && categories.second <  fDictCategoriesStart) ||
+                (categories.first >= fDictCategoriesStart && categories.second >= fDictCategoriesStart));
         for (RangeDescriptor rd = fRangeList; rd != null; rd = rd.fNext) {
-            int rangeNum = rd.fNum & ~DICT_BIT;
-            int rangeDict = rd.fNum & DICT_BIT;
+            int rangeNum = rd.fNum;
             if (rangeNum == categories.second) {
-                rd.fNum = categories.first | rangeDict;
+                rd.fNum = categories.first;
             } else if (rangeNum > categories.second) {
                 rd.fNum--;
             }
         }
         --fGroupCount;
+        if (fDictCategoriesStart >= categories.second) {
```

Comment:
optional: turn around as suggested in C++

Reply:
Done